common/dev/peci.c

```diff
@@ -0,0 +1,268 @@
+		    (sensor_config[sensor_config_index_map[SENSOR_NUM_TEMP_CPU_MARGIN]]
+			     .cache_status == SENSOR_READ_SUCCESS)) {
+			*reading =
+				sensor_config[sensor_config_index_map[SENSOR_NUM_TEMP_CPU_TJMAX]]
+					.cache -
+				sensor_config[sensor_config_index_map[SENSOR_NUM_TEMP_CPU_MARGIN]]
+					.cache;
+			cfg->cache = *reading;
+			cfg->cache_status = SENSOR_READ_SUCCESS;
+			return true;
+		} else {
+			cfg->cache_status = SENSOR_FAIL_TO_ACCESS;
+			return false;
+		}
+	} else if (sensor_num == SENSOR_NUM_PWR_CPU) {
+		ret = peci_get_power(sensor_num, &val);
+		if (!ret) {
+			cfg->cache_status = SENSOR_FAIL_TO_ACCESS;
+			return false;
+		}
+		*reading = (float)val;
+		cfg->cache = *reading;
+		cfg->cache_status = SENSOR_READ_SUCCESS;
+		return true;
+	} else if ((sensor_num == SENSOR_NUM_TEMP_CPU_TJMAX) &&
+		   (cfg->cache_status == SENSOR_READ_SUCCESS)) { // Tjmax only need to get once
+		*reading = cfg->cache;
+		return true;
+	} else { // The other sensors
+		if (cfg->port == NONE) {
+			cfg->cache_status = SENSOR_NOT_FOUND;
+			return false;
+		}
+
+		read_buf = (uint8_t *)malloc(PECI_READ_LEN * sizeof(uint8_t));
+		if (read_buf == NULL) {
+			printf("PECI buffer alloc fail\n");
+			return false;
+		}
+
+		uint8_t *complete_code = &read_buf[0];
+		int retry_count = PECI_RETRY_MAX;
+		while (1) {
+			ret = peci_read(cmd, cfg->port, cfg->target_addr, cfg->offset, PECI_READ_LEN,
+					read_buf);
+			if (ret == 0) {
+				if (*complete_code == PECI_CC_RSP_SUCCESS) {
+					break;
+				}
+			} else if (--retry_count) { // wait for 10ms before next retry
+				k_msleep(10);
+			} else { // all retries failed exit the loop
+				break;
+			}
+		}
+		if (retry_count == 0) {
+			if (ret) {
+				printf("PECI sensor [0x%x] retry read fail, retry time: %d, completion_code: 0x%x\n",
+				       sensor_num, PECI_RETRY_MAX, *complete_code);
+				cfg->cache_status = SENSOR_FAIL_TO_ACCESS;
+			} else {
+				switch (*complete_code) {
+				case PECI_CC_RSP_TIMEOUT:
+				case PECI_CC_OUT_OF_RESOURCES_TIMEOUT:
+					printf("PECI sensor [0x%x] response timeout\n", sensor_num);
+					cfg->cache_status = SENSOR_FAIL_TO_ACCESS;
+					break;
+				case PECI_CC_ILLEGAL_REQUEST:
+					printf("PECI sensor [0x%x] unknown request\n", sensor_num);
+					cfg->cache_status = SENSOR_NOT_FOUND;
+					break;
+				default:
+					printf("PECI sensor [0x%x] control hardware, firmware or associated logic error\n",
+					       sensor_num);
+					cfg->cache_status = SENSOR_UNSPECIFIED_ERROR;
+					break;
+				}
+			}
+
+			free(read_buf);
+			return false;
+		}
+
+		// PECI_CC_RSP_SUCCESS
+		if (sensor_num == SENSOR_NUM_TEMP_CPU_MARGIN) {
+			val = ((0xFFFF - ((read_buf[2] << 8) | read_buf[1])) >> 6) + 1;
+		} else if (sensor_num == SENSOR_NUM_TEMP_CPU_TJMAX) {
+			val = read_buf[3];
+		} else {
+			val = read_buf[1];
+		}
+
+		*reading = (float)val;
+		cfg->cache = *reading;
+		cfg->cache_status = SENSOR_READ_SUCCESS;
+		free(read_buf);
```

Comment:
SAFE_FREE

Reply:
The optimization code is based on v2022.07.01. After that, the main branch has been phased in some
commits related to sensor monitoring. Those changes will be phased in main_v2 in the other PRs. This
problem will be solved in those PRs.
Please check #223  .